plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -625,18 +625,22 @@ enterComputeCek = computeCek (toWordArray 0) where
         returnCek unbudgetedSteps' ctx (VCon val)
     computeCek !unbudgetedSteps !ctx !env (LamAbs _ name body) = do
         !unbudgetedSteps' <- stepAndMaybeSpend BLamAbs unbudgetedSteps
-        returnCek unbudgetedSteps' ctx (VLamAbs name body env)
+        case ctx of
+            FrameAwaitFun arg ctx' -> applyLam unbudgetedSteps ctx' env body arg
+            _                      -> returnCek unbudgetedSteps' ctx (VLamAbs name body env)
     computeCek !unbudgetedSteps !ctx !env (Delay _ body) = do
         !unbudgetedSteps' <- stepAndMaybeSpend BDelay unbudgetedSteps
-        returnCek unbudgetedSteps' ctx (VDelay body env)
+        case ctx of
+            FrameForce ctx' -> forceDelay unbudgetedSteps ctx' env body
+            _               -> returnCek unbudgetedSteps' ctx (VDelay body env)
```

Comment:
The only other possible alternatives are `NoFrame` and `FrameAwaitArg`, right? Can we then handle those here as well and throw on `FrameApplyFun`, thus cutting one pattern match away across the board for every `VDelay` object?

Does some similar logic apply to `VLamAbs`?

It's probably indeed quite some duplication, but maybe we can unfold `returnCek` a bit by making it receive `returnCek` and `computeCek` and whatever else as additional arguments, so that `returnCek` isn't perceived as recursive by GHC and then tie the knot ourselves? It'll probably break join-point generation, though.

Reply:
Hmm, does `inline` work on recursive functions? I wonder if I could just call `inline` at the uses of `returnCek`...

Reply:
It works! And looking at the Core it does exactly what you asked for. I'll benchmark it.

Reply:
Weirdly, this version is slightly worse???

Reply:
> It works!

You kidding me???

OK, after the initial shock I think I've found the answer: not all functions from a mutually recursive family are non-inlinable, only the [loop-breaker](https://wiki.haskell.org/Inlining_and_Specialisation#What_is_a_loop-breaker.3F) is.

> Weirdly, this version is slightly worse???

Maybe benchmarking is being weird?

Reply:
Not sure, I ran both again and it was still worse? Maybe the first time was a fluke?